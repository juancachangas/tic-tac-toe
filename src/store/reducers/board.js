import { SET_MOVE, SET_NEW_GAME, SET_STATUS } from '../actions/board';


export const INITIAL_STATE = {
  fields: [null, null, null,
    null, null, null,
    null, null, null,
  ],
  turn: '0',
  status: 'new',
};

export const board = (state = INITIAL_STATE, action) => {
  const { payload, type } = action;
  const { fields, turn } = state;
  switch (type) {
    case SET_MOVE: {
      const newFields = fields.map((field, index) => (
        index === payload.position
          ? turn
          : field));
      const isEmptyPosition = fields[payload.position] === null;
      if (isEmptyPosition) {
        return {
          status: 'playing',
          turn: turn === '0' ? '1' : '0',
          fields: newFields,
        };
      }
      return {
        ...state,
        status: 'wrong_move',
      };
    }
    case SET_STATUS: {
      return {
        ...state,
        status: payload.status,
      };
    }
    case SET_NEW_GAME: {
      return { ...INITIAL_STATE };
    }
    default: {
      return {
        ...state,
      };
    }
  }
};
