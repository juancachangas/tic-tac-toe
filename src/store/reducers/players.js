import { ADD_PLAYER } from '../actions/players';


export const INITIAL_STATE = [];

export const players = (state = INITIAL_STATE, action) => {
  const { payload, type } = action;
  switch (type) {
    case ADD_PLAYER: {
      const newPlayer = payload.player;
      const newPlayers = state.length === 2 ? [...state] : [...state, newPlayer];
      return newPlayers;
    }
    default:
      return [...state];
  }
};
