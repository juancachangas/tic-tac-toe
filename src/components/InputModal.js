/*
 * You do not need to separate components and containers.
 * Feel free to connect your state and components directly in here (or other components).
 */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addPlayer as setPlayer } from '../store/actions/players';

class InputModal extends Component {
  constructor() {
    super();
    this.state = {
      inputValue: '',
    };
  }

  onChange = (ev) => {
    const text = ev.target.value;
    this.setState({ inputValue: text });
  }

  render() {
    const { inputValue } = this.state;
    const { players, addPlayer } = this.props;
    return (
      <article className={`Modal winner-${players.length}`}>
        <label htmlFor="player-name" id="player-name-text">
          Player name:
          <input id="player-name" type="text" value={inputValue} onChange={this.onChange} required />
        </label>
        <button
          type="button"
          onClick={() => {
            addPlayer(inputValue);
            this.setState({ inputValue: '' });
          }}
          disabled={inputValue.trim() === ''}
        >
          Add +
        </button>
      </article>
    );
  }
}
export const mapStateToProps = (state) => {
  const { players = [] } = state;
  return {
    players,
  };
};
export const mapDispatchToProps = dispatch => ({
  addPlayer: player => dispatch(setPlayer(player)),
});

export const PlayerModal = connect(
  mapStateToProps,
  mapDispatchToProps,
)(InputModal);
