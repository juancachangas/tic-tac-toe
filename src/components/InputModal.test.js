import React from 'react';
import { shallow } from 'enzyme';

import InputModal, { mapStateToProps, mapDispatchToProps } from './InputModal';
import { INITIAL_STATE } from '../store/reducers/players';

describe('InputModal component is', () => {
  it('rendering default', () => {
    const wrapper = shallow(
      <InputModal players={[]} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('rendering seccondPlayer', () => {
    const wrapper = shallow(
      <InputModal players={['player1']} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
describe('mapStateToProps', () => {
  it('returns proper props from state', () => {
    const state = {
      players: INITIAL_STATE,
    };
    const expectedProps = {
      players: [],
    };
    expect(mapStateToProps(state)).toMatchObject(expectedProps);
  });
});
describe('mapDispatchToProps', () => {
  it('mapDispatchToProps returns required functions', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);

    expect(typeof props.addPlayer).toBeDefined();
    expect(typeof props.addPlayer).toEqual('function');
  });
});
