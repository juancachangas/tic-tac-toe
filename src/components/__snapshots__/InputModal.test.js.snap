// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputModal component is rendering default 1`] = `
<article
  className="Modal winner-0"
>
  <label
    htmlFor="player-name"
    id="player-name-text"
  >
    Player name:
    <input
      id="player-name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </label>
  <button
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Add +
  </button>
</article>
`;

exports[`InputModal component is rendering secondPlayer 1`] = `
<article
  className="Modal winner-1"
>
  <label
    htmlFor="player-name"
    id="player-name-text"
  >
    Player name:
    <input
      id="player-name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </label>
  <button
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Add +
  </button>
</article>
`;
